test/state/mpt_hash.cpp

```diff
@@ -0,0 +1,36 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2022 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "mpt_hash.hpp"
+#include "account.hpp"
+#include "mpt.hpp"
+#include "rlp.hpp"
+
+namespace evmone::state
+{
+namespace
+{
+hash256 mpt_hash(const std::unordered_map<hash256, StorageValue>& storage)
+{
+    MPT trie;
+    for (const auto& [key, value] : storage)
+    {
+        if (!is_zero(value.current))  // Skip "deleted" values.
```

Comment:
Maybe this is rather logic of how MPT data structure is used in Ethereum... so not logic of hashing itself, may be better kept outside.

But I don't mind either way.

Reply:
Both "hash functions" are use-case specific because at least you need to decide how to serialize a structure. As long as we will keep original and current value together we cannot erase instead zeroing.

In more generic fashion in this case `std::copy_if` could be used if we support `std::insterter()`. In the other case we would need something like "transform if" so ranges should be more suited. Interesting exercise, but decided is too big detour.